Client/Frontend/Browser/BrowserViewController.swift

```diff
@@ -847,6 +847,48 @@ extension BrowserViewController: BrowserToolbarDelegate {
     }
 }
 
+protocol WindowCloseHelperDelegate: class {
+    func windowCloseHelper(windowCloseHelper: WindowCloseHelper, didRequestToCloseBrowser browser: Browser)
+}
+
+class WindowCloseHelper: BrowserHelper {
+    weak var delegate: WindowCloseHelperDelegate?
+    private weak var browser: Browser?
+
+    required init(browser: Browser) {
+        self.browser = browser
+        if let path = NSBundle.mainBundle().pathForResource("WindowCloseHelper", ofType: "js") {
+            if let source = NSString(contentsOfFile: path, encoding: NSUTF8StringEncoding, error: nil) as? String {
+                var userScript = WKUserScript(source: source, injectionTime: WKUserScriptInjectionTime.AtDocumentEnd, forMainFrameOnly: true)
+                browser.webView!.configuration.userContentController.addUserScript(userScript)
+            }
+        }
+    }
+
+    func scriptMessageHandlerName() -> String? {
+        return "windowCloseHelper"
+    }
+
+    func userContentController(userContentController: WKUserContentController, didReceiveScriptMessage message: WKScriptMessage) {
+        println("This is WindowCloseHelper.userContentController(_:, didReceiveScriptMessage:)")
+        if let browser = browser {
+            dispatch_async(dispatch_get_main_queue()) {
```

Comment:
I think it is good practice for delegates, for which we do not know on which thread they run, to call back on the main thread. Then the caller does not have to worry about it. Should probably wrap it in a check to see if we are already on the main thread or not.